### PR TITLE
Feature/time awareness

### PR DESCRIPTION
**Pull request by 0xangelo**
First part of #64

### CODE REVIEW

File: mapo/envs/navigation.py [hunk 1]
```diff
@@ -110,11 +110,11 @@ def _transition_log_prob_fn(self, state, action, next_state):
     def _reward_fn(self, state, action, next_state):
         # pylint: disable=invalid-unary-operand-type
         goal = tf.constant(self._end, name="goal")
-        return -tf.norm(state - goal, axis=-1)
+        return -tf.norm(next_state - goal, axis=-1)
```
Comment (0xangelo) [inline mapo/envs/navigation.py:113]: Need to condition on the next state in order to reach the goal in 20 steps. Furthermore, by conditioning on the current state we would always receive the same penalty in the first time step, regardless of the agent's decision.

File: mapo/envs/tests/test_navigation.py [hunk 2]
```diff
@@ -144,21 +133,19 @@ def test_transition(env):
...
     with env._graph.as_default():
         dummy_action = env.action_space.sample()
-        dummy_next_state = env.observation_space.sample()
+        dummy_state = env.observation_space.sample()
 
-        reward = env._reward(env._end, dummy_action, dummy_next_state)
+        reward = env._reward(dummy_state, dummy_action, env._end)
         assert np.allclose(reward, 0.0)
-        reward = env._reward(env._start, dummy_action, dummy_next_state)
+        reward = env._reward(dummy_state, dummy_action, env._start)
```
Comment (0xangelo) [inline mapo/envs/tests/test_navigation.py:143]: Imagine the agent arrives at the goal in the second to last time step. Then, even if the agent moves away from the goal in the final step, we would be rewarding that behavior (the transition would have high reward even though it is a bad one). This is another motivation for the change in the reward function.

File: mapo/envs/navigation.py (hunk 1 above)
Comment (thiagopbueno) [reply mapo/envs/navigation.py:113]: This is a bit controversial from the perspective of the Bellman equation, but I don't see any problem about using the `next_state` here.

File: mapo/envs/__init__.py [hunk 3]
```diff
@@ -56,10 +56,10 @@ def _build_ops(self):
     def step(self, action):
         next_state, _ = self._transition(self._state, action)
         reward = self._reward(self._state, action, next_state)
-        done = self._terminal()
-        info = self._info()
         self._state = next_state
         self._timestep += 1
+        done = self._terminal()
```
Comment (thiagopbueno) [inline mapo/envs/__init__.py:61]: Thanks for debugging this.

File: mapo/envs/navigation.py [hunk 4]
```diff
@@ -110,11 +110,11 @@ def _transition_log_prob_fn(self, state, action, next_state):
...
 
     def _terminal(self):
         reached_goal = np.allclose(self._state, self._end, atol=1e-1)
-        timeout = self._timestep > self._horizon
+        timeout = self._timestep >= self._horizon
```
Comment (thiagopbueno) [inline mapo/envs/navigation.py:117]: My bad.

File: mapo/envs/tests/test_navigation.py [hunk 5]
```diff
@@ -17,13 +16,11 @@
...
 
 
-def get_gym_envs():
-    env0 = _import_navigation_v0(None)
-    env1 = _import_navigation_v1(None)
-    envs = [env0, env1]
-    for env in envs:
-        env.reset()
-    return envs
+@pytest.fixture(params=[_import_navigation_v0, _import_navigation_v1])
```
Comment (thiagopbueno) [inline mapo/envs/tests/test_navigation.py:19]: Good trick. Thanks for the nice refactoring.

File: mapo/envs/tests/test_navigation.py (hunk 2 above)
Comment (thiagopbueno) [reply mapo/envs/tests/test_navigation.py:143]: If the agent arrives at the goal (or close by some metric), then there's no further steps as the episode will terminate.

File: mapo/envs/tests/test_navigation.py [hunk 6]
```diff
@@ -170,11 +157,20 @@ def test_terminal(env):
         assert env._terminal()
 
     env.reset()
-    env._timestep = env._horizon + 1
+    env._timestep = env._horizon
     assert env._terminal()
 
 
-@pytest.mark.parametrize("env", get_gym_envs())
+def test_rollout_length(env):
```
Comment (thiagopbueno) [inline mapo/envs/tests/test_navigation.py:164]: Nice simple test.

**Review by thiagopbueno — APPROVED** [review]

**Pull request merged**